File: day03/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define MAX_LEN 256
#define BIT_COUNT 5

int binToDec(char *bin) {
  int result = 0;
  while(*bin) {
    result *= 2;
    result += *bin == '1' ? 1 : 0;
    ++bin;
  }
  return result;
}

bool startsWith(char *haystack, char *needle, int curBitCount) {
  printf("\t\t%s %s\n", haystack, needle);
  for (int i = 0; i < curBitCount; i++) {
    if (haystack[i] != needle[i]) {
      return false;
    }
  }
  return true;
}

int singleIn(char *value, char* values[], int curBitCount) {
  int index = -1;
  for(int i = 0; i < 12; i++) {
    printf("\t%s, index = %d / %s\n", values[i], i, value);
    if(startsWith(values[i], value, curBitCount)) {
      if(index != -1) {
        printf("\t!!! breaking\n");
        return -1;
      }
      index = i;
    }
  }
  return index;
}

int main(int argc, char *argv[]) {
  FILE* file;
  char filename[] = "input.txt";
  if(argc > 1) {
    strcpy(filename, argv[1]);
  }
  file = fopen(filename, "r");
  if(file == NULL) {
    perror("failed to read file");
    return 1;
  }

  char buffer[MAX_LEN];
  int counter = 0, i;
  int part1 = 0, oneCount[BIT_COUNT] = {};
  char gamma[BIT_COUNT + 1] = {}, epsilon[BIT_COUNT + 1] = {};
  int part2 = 0;
  char co2[BIT_COUNT + 1] = {}, oxygen[BIT_COUNT + 1] = {}, *values[12] = {};

  for(i = 0; i < BIT_COUNT; i++) {
    oneCount[i] = 0;
  }

  while(fgets(buffer, MAX_LEN, file)) {
    buffer[strcspn(buffer, "\n")] = 0;
    values[counter] = buffer;
    printf("%s\n", buffer);
    for(i = 0; i < BIT_COUNT; i++) {
      if(buffer[i] == '1') {
        oneCount[i] += 1;
      }
    }

    counter++;
  }
  printf("\n");

  for(i = 0; i < BIT_COUNT; i++) {
    // printf("%d ", oneCount[i]);
    if(oneCount[i] > counter / 2) {
      gamma[i] = '1';
      epsilon[i] = '0';
      co2[i] = '0';
      oxygen[i] = '1';
    } else {
      gamma[i] = '0';
      epsilon[i] = '1';
      co2[i] = '1';
      oxygen[i] = '0';
    }
    
    printf("%s len(%d)\n", co2, i + 1);
    int co2SingleIn = singleIn(co2, values, i + 1);
    int oxygenSingleIn = singleIn(oxygen, values, i + 1);
    if(co2SingleIn != -1) {
      printf("\t%s : %d\t", co2, co2SingleIn);
      *co2 = *values[co2SingleIn];
    }
    if(oxygenSingleIn != -1) {
      *oxygen = *values[oxygenSingleIn];
    }
  }

  printf("\n");
  gamma[BIT_COUNT] = 0;
  epsilon[BIT_COUNT] = 0;

  fclose(file);
  part1 = binToDec(gamma) * binToDec(epsilon);
  part2 = binToDec(co2) * binToDec(oxygen);
  printf("counter: %d\n", counter);
  printf("Part 1: %s * %s = %d\n", gamma, epsilon, part1);
  printf("Part 2: %s * %s = %d\n", co2, oxygen, part2);

  return 0;
}

File: day03/main.dart
import 'dart:async';
import 'dart:io';
import 'dart:convert';

Future<List<String>> getLines(filename) async {
  List<String> lines = [];
  await new File(filename)
      .openRead()
      .map(utf8.decode)
      .transform(new LineSplitter())
      .forEach((line) {
    lines.add(line);
  });
  return lines;
}

void main(List<String> arguments) async {
  List<String> lines = [];
  String filename = 'input.txt';
  int bitcount = 12;
  if (arguments.length == 1) {
    filename = arguments[0];
  }

  List<int> oneCount = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
  int count;
  String gamma = '', epsilon = '';
  List<String> co2 = [], oxygen = [];
  int part1 = 0;
  int part2 = 0;

  lines = await getLines(filename);
  count = lines.length;
  for (final line in lines) {
    bitcount = line.length;
    for (int i = 0; i < bitcount; i++) {
      if (line.substring(i, i + 1) == '1') {
        oneCount[i] += 1;
      }
    }
  }
  for (int i = 0; i < bitcount; i++) {
    if (oneCount[i] > count / 2) {
      gamma += '1';
      epsilon += '0';
    } else {
      gamma += '0';
      epsilon += '1';
    }
  }

  oxygen = [...lines]; // fancy spread operator for copying
  co2 = [...lines];
  int oneCountOxygen = 0;
  int zeroCountCo2 = 0;

  for (int i = 0; oxygen.length > 1; i++) {
    oneCountOxygen = 0;
    for (final data in oxygen) {
      if (data.substring(i, i + 1) == '1') {
        oneCountOxygen += 1;
      }
    }
    if (oneCountOxygen >= oxygen.length - oneCountOxygen) {
      oxygen.removeWhere((element) => element.substring(i, i + 1) == '0');
    } else {
      oxygen.removeWhere((element) => element.substring(i, i + 1) == '1');
    }
  }

  for (int i = 0; co2.length > 1; i++) {
    zeroCountCo2 = 0;
    for (final data in co2) {
      if (data.substring(i, i + 1) == '0') {
        zeroCountCo2 += 1;
      }
    }
    if (zeroCountCo2 <= co2.length - zeroCountCo2) {
      co2.removeWhere((element) => element.substring(i, i + 1) == '1');
    } else {
      co2.removeWhere((element) => element.substring(i, i + 1) == '0');
    }
  }

  part1 = int.parse(gamma, radix: 2) * int.parse(epsilon, radix: 2);
  part2 = int.parse(co2[0], radix: 2) * int.parse(oxygen[0], radix: 2);
  print('part 1: $part1');
  print('part 2: $part2');
}
